### ISSUE TITLE
Dotted borders render incorrectly

### ISSUE THREAD

**Issue by nc4rrillo** [servo/webrender#1932]
Hi,
I've noticed dotted borders render the corners seemingly incorrectly.

![](https://i.imgur.com/6x0xBa0.png)

```
---
root: 
  items: 
    - 
      bounds: [5, 5, 490, 490]
      "clip-rect": [5, 5, 490, 490]
      "clip-and-scroll": 0
      "backface-visible": true
      type: border
      width: 25
      "border-type": normal
      color: black
      style: dotted
  id: [0, 0]
pipelines: []
```

**Comment by nc4rrillo** [comment]
Seems related to having a radius < width for the border

```
---
root: 
  items: 
    - 
      bounds: [5, 5, 490, 490]
      "clip-rect": [5, 5, 490, 490]
      "clip-and-scroll": 0
      "backface-visible": true
      type: border
      width: 25
      "border-type": normal
      color: black
      style: dotted
      "radius": 35
  id: [0, 0]
pipelines: []
```

renders as: 


![](https://i.imgur.com/dylVAuL.png)

**Comment by nical** [comment]
I traced it down to the creation of `BorderCornerClipSource` in border.rs where we compute the ellipse of a dotted corner with negative radii due to subtracting the width to the corner radius (good catch @nc4rrillo!). The code that positions the dots for the corner expects positive ellipse raddii, and using the abs fixes the missing pieces.
However, because of the way this code places the dots on the corners and edges, we end up with overlapping dots in some cases which is somewhat underwhelming. Is there any bit of specification about how the dots should be positioned?

All I can find in the CSS spec is "The border is a series of dots."

**Comment by pcwalton** [comment]
@nical Based on experience with CSS 2.1, I highly doubt there's a detailed spec for dotted borders.

**Comment by nical** [comment]
bugzilla entry: https://bugzilla.mozilla.org/show_bug.cgi?id=1415246

**Comment by nical** [comment]
The clipped corners are fixed in #2007, the remaining work (ugly overlap) is tracked in #2013.
